ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs

```diff
@@ -776,137 +789,131 @@ newtype Completion m peeraddr peerconn =
-     | TraceDemoteWarmDone     Int Int peeraddr
+     | TraceDemoteWarmDone     !Int !Int !peeraddr
 
      -- | target established big ledger peers, actual established big ledger
      -- peers, selected peers
-     | TraceDemoteWarmBigLedgerPeers    Int Int (Set peeraddr)
+     | TraceDemoteWarmBigLedgerPeers    !Int !Int !(Set peeraddr)
      -- | target established big ledger peers, actual established big ledger
      -- peers, peer, reason
-     | TraceDemoteWarmBigLedgerPeerFailed   Int Int peeraddr SomeException
+     | TraceDemoteWarmBigLedgerPeerFailed   !Int !Int !peeraddr !SomeException
      -- | target established big ledger peers, actual established big ledger
      -- peers, peer
-     | TraceDemoteWarmBigLedgerPeerDone     Int Int peeraddr
+     | TraceDemoteWarmBigLedgerPeerDone     !Int !Int !peeraddr
 
      -- | target active, actual active, selected peers
-     | TraceDemoteHotPeers     Int Int (Set peeraddr)
+     | TraceDemoteHotPeers     !Int !Int !(Set peeraddr)
      -- | local per-group (target active, actual active), selected peers
-     | TraceDemoteLocalHotPeers [(HotValency, Int)] (Set peeraddr)
+     | TraceDemoteLocalHotPeers ![(HotValency, Int)] !(Set peeraddr)
      -- | target active, actual active, peer, reason
-     | TraceDemoteHotFailed    Int Int peeraddr SomeException
+     | TraceDemoteHotFailed    !Int !Int !peeraddr !SomeException
      -- | target active, actual active, peer
-     | TraceDemoteHotDone      Int Int peeraddr
+     | TraceDemoteHotDone      !Int !Int !peeraddr
 
      -- | target active big ledger peers, actual active big ledger peers,
      -- selected peers
-     | TraceDemoteHotBigLedgerPeers      Int Int (Set peeraddr)
+     | TraceDemoteHotBigLedgerPeers      !Int !Int !(Set peeraddr)
      -- | target active big ledger peers, actual active big ledger peers, peer,
      -- reason
-     | TraceDemoteHotBigLedgerPeerFailed Int Int peeraddr SomeException
+     | TraceDemoteHotBigLedgerPeerFailed !Int !Int !peeraddr !SomeException
      -- | target active big ledger peers, actual active big ledger peers, peer
-     | TraceDemoteHotBigLedgerPeerDone   Int Int peeraddr
+     | TraceDemoteHotBigLedgerPeerDone   !Int !Int !peeraddr
 
-     | TraceDemoteAsynchronous      (Map peeraddr (PeerStatus, Maybe ReconnectDelay))
-     | TraceDemoteLocalAsynchronous (Map peeraddr (PeerStatus, Maybe ReconnectDelay))
+     | TraceDemoteAsynchronous      !(Map peeraddr (PeerStatus, Maybe ReconnectDelay))
+     | TraceDemoteLocalAsynchronous !(Map peeraddr (PeerStatus, Maybe ReconnectDelay))
      | TraceDemoteBigLedgerPeersAsynchronous
-                                    (Map peeraddr (PeerStatus, Maybe ReconnectDelay))
+                                    !(Map peeraddr (PeerStatus, Maybe ReconnectDelay))
 
      | TraceGovernorWakeup
-     | TraceChurnWait          DiffTime
-     | TraceChurnMode          ChurnMode
+     | TraceChurnWait          !DiffTime
+     | TraceChurnMode          !ChurnMode
```

Comment:
The tracing team prefers us to keep the traces lazy, as quite often they are thrown away.  Does this actually improves the performance of the tests?

Reply:
I suspect it doesn't, unless we are retaining the whole trace somewhere, what actually be a bigger issue; but if we are analysing the simulation like a stream (visiting each trace message, and throwing it away so it's GC-ed, then strictness annotations should be unnecessary - each test focuses only on some trace messages and the rest doesn't need to be evaluated (unless there are dependencies between different trace messages, which would be a bit surprising).

Reply:
Ah got it! It is an artifact of some experiments I was doing, I will delete that!